lib/fob/fragment_builder.ex

```diff
@@ -16,9 +16,39 @@ defmodule Fob.FragmentBuilder do
     end
   end
 
-  defmacro build_from_existing(binding \\ [], expr, params) do
-    build(binding, expr, params, __CALLER__)
+  defmacro build_from_existing(binding \\ [], expr) do
+    build(binding, expr, [], __CALLER__)
   end
 
   # chaps-ignore-start
+
+  def table_for_fragment({:fragment, [], _} = frag) do
+    Macro.prewalk(frag, nil, fn ast, acc ->
+      case ast do
+        {:&, [], [table_idx]} when is_integer(table_idx) ->
+          {ast, table_idx}
+
+        _ ->
+          {ast, acc}
+      end
+    end)
+    |> elem(1)
+  end
+
+  def column_for_query_fragment(
+        {:fragment, [], _} = frag,
+        %Ecto.Query{} = query
+      ) do
+    query.select.expr
+    |> Macro.prewalk(nil, fn ast, acc ->
+      case ast do
+        {virtual_column, ^frag} when is_atom(virtual_column) ->
+          {ast, virtual_column}
+
+        _ ->
+          {ast, acc}
+      end
+    end)
+    |> elem(1)
+  end
```

Comment:
nice, this makes sense 👍

Reply:
Nice, I wasn't sure if this was reasonable code, it's correct but it is also an odd use of `prewalk`, If we bumped the elixir version we could use `prewalker` which would read much nicer and do the proper halting of the ast. But it's probably best to keep it as is for now.